### PR TITLE
Neuralnet

### PR DESCRIPTION
**Pull request by mhw32**
- 99% accuracy. 
- finished linking all parts of disambiguator.

### CODE REVIEW

File: disambiguator.py [hunk 1]
```diff
@@ -96,6 +93,7 @@ def init_prior_pos_proba(
     cat_lookup = group_categories()
     num_tags = descriptor_array.shape[0]
     tag_counts = defaultdict(lambda: np.zeros(num_tags))
+    tag_counts = defaultdict(lambda: np.zeros(num_tags))
```
Comment (kennyhlam) [inline disambiguator.py]: lol, remove

File: disambiguator.py [hunk 2]
```diff
@@ -311,20 +325,20 @@ def group_categories():
...
     mapper["HVZ"] = [VERB]                        # has
     mapper["IN"] = [PREPOSITION]                  # preposition
     mapper["JJ"] = [MODIFIER]                     # adjective
-    mapper["JJR"] = [MODIFIER]                    # comparative adjective 
+    mapper["JJR"] = [MODIFIER]                    # comparative adjective
     mapper["JJS"] = [MODIFIER]                    # semantically superlative adjective [chief, top]
     mapper["JJT"] = [MODIFIER]                    # morphologically superlative adjective [ biggest ]
     mapper["MD"] = [OTHERS]                       # modal auxiliary [can, should, will]
-    mapper["NC"] = None                           # cited word (hypenated after regular tag)
+    mapper["NC"] = [ARTICLE]                      # cited word (hypenated after regular tag)
```
Comment (kennyhlam) [inline disambiguator.py]: this isn't an article right?

File: create_toy_data.py [hunk 3]
```diff
@@ -62,42 +65,115 @@ def load_data(
...
-                yield((cur_doc_word, cur_doc_pos))
-            else:
-                yield cur_doc_word
+    if not out_data is None:
+        tr_out_data, te_out_data = out_data[tr_idx, :], out_data[te_idx, :]
+        return (tr_in_data, te_in_data), (tr_out_data, te_out_data)
 
+    return (tr_in_data, te_in_data)
+
+def get_auc(outputs, probas):
```
Comment (kennyhlam) [inline create_toy_data.py]: does this really belong in create_toy_data? it's just some auxillary function to calculate the goodness of your probabilities

File: disambiguator.py [hunk 4]
```diff
@@ -371,95 +385,143 @@ def group_categories():
...
 
-def accuracy(params, inputs, targets):
-    target_class = np.argmax(targets, axis=1)
-    predicted_class = np.argmax(neural_net_predict(params, inputs), axis=1)
-    return np.mean(predicted_class == target_class)
+    target_idx = np.where(tag_order == target_tag)[0][0]
+    eos_idx = np.where(darrays[:,target_idx] > 0)[0]
 
+    new_darrays = np.zeros((eos_idx.shape[0],darrays.shape[1]*(2*num_grams + 1)))
+    new_labels = np.zeros((eos_idx.shape[0], labels.shape[1]))
```
Comment (kennyhlam) [inline disambiguator.py:451]: I'm not sure what you're going for here. `labels` is a 1-D array when you call it from `create_features_labels`, so `.shape[1]` fails

File: disambiguator.py [hunk 5]
```diff
@@ -371,95 +385,143 @@ def group_categories():
...
-    return optimized_params
+        num_epochs=20, step_size=0.001, L2_reg=1.0)
+
+    # save the weights
+    np.save('storage/trained_weights.npy', trained_weights)
+
+    y_pred = nn.neural_net_predict(trained_weights, X_test)
+    # don't forget to exp
+    print("auc: {}".format(
+        ctd.get_auc(y_test[:, 1], np.exp(y_pred))))
```
Comment (kennyhlam) [inline disambiguator.py]: Why's it `y_test[:, 1]`, shouldn't it be 0? Isn't it just a 1-D array? This depends on what you were hoping to go for in the `make_grams`, so could be a no-op

**Review by kennyhlam — CHANGES_REQUESTED** [review]
just some questions, and the `make_grams` fails for me

**Comment by kennyhlam** [comment]
Also, we're slowly but surely building more files that are going into `storage/`, based on the number of parameters we're having to store. For now, it's still pretty easy and manageable to generate them, but I'm wondering if we should start a `datagen` folder with binaries (or one really big one) to generate all our parameters in turn. I predict we'll have an increasing number of parameters we'll want to store in the future, especially if we try other techniques.

The generation of our parameters also requires a strict ordering (must write/load tag order before making k-grams around EOS punctuation). I've personally just been shoving stuff the steps into `workspace.py`, just cause it's there, but I don't know what you're thinking going forward.

File: create_toy_data.py [hunk 6]
```diff
@@ -62,42 +65,115 @@ def load_data(
...
+                            token repr. EOS.
         Returns
         -------
         data : Python generator
                yields a single token (w/ POS)
     '''
 
+    default_splitter = nltk.data.load(
+        'tokenizers/punkt/english.pickle')
+
```
Comment (kennyhlam) [inline create_toy_data.py:87]: So by switching to the punkt tokenizer, we're getting more "bad splits":

reading: datasets/brown/raw/cl11
found an item with no tag: M.E.

reading: datasets/brown/raw/cj24
found an item with no tag: B.
found an item with no tag: B.

reading: datasets/brown/raw/cj58
found an item with no tag: B.

reading: datasets/brown/raw/cj73
found an item with no tag: '?

reading: datasets/brown/raw/cf29
found an item with no tag: TR.

reading: datasets/brown/raw/ca15
found an item with no tag: U.

reading: datasets/brown/raw/cd03
found an item with no tag: B.B.C.

reading: datasets/brown/raw/ca40
found an item with no tag: A.A.U.

**Comment by kennyhlam** [comment]
I looked at the neuralnet stuff, I don't know what the exact math that's supposed to happen is, but I read all the code and it looks reasonable.

I made https://github.com/mhw32/Adaware/pull/7

If you're good with that, I'm good to merge everything into master.

File: disambiguator.py (hunk 1 above)
Comment (mhw32) [reply disambiguator.py]: deleted.

File: disambiguator.py (hunk 2 above)
Comment (mhw32) [reply disambiguator.py]: Added as [OTHERS]

File: disambiguator.py (hunk 4 above)
Comment (mhw32) [reply disambiguator.py:451]: added back in the get_dummies(...) call

File: disambiguator.py (hunk 5 above)
Comment (mhw32) [reply disambiguator.py]: added back in the get_dummies(...) call

File: create_toy_data.py (hunk 3 above)
Comment (mhw32) [reply create_toy_data.py]: moved to separate file called "metrics"

File: create_toy_data.py (hunk 6 above)
Comment (mhw32) [reply create_toy_data.py:87]: Hmm... well what this tell me is that we need a better labeling system... Maybe we can find a dataset somewhere

**Comment by mhw32** [comment]
Pushed simple changes. Let me know if we are good to go. @kennyhlam

**Review by kennyhlam — APPROVED** [review]
